model/nh_utils.F90

```diff
@@ -0,0 +1,1916 @@
+   real, intent(in), dimension(is-ng:ie+ng,js-ng:je+ng,km):: w3
+! OUTPUT PARAMETERS
+   real, intent(inout), dimension(is-ng:ie+ng,js-ng:je+ng,km+1):: gz
+   real, intent(  out), dimension(is-ng:ie+ng,js-ng:je+ng,km+1):: pef
+! Local:
+  real, dimension(is-1:ie+1,km  ):: dm, dz2, w2, pm2, gm2, cp2
+  real, dimension(is-1:ie+1,km+1):: pem, pe2, peg
+  real gama, rgrav
+  integer i, j, k
+  integer is1, ie1
+
+    gama = 1./(1.-akap)
+   rgrav = 1./grav
+
+   is1 = is - 1
+   ie1 = ie + 1
+
+!$OMP parallel do default(none) shared(js,je,is1,ie1,km,delp,pef,ptop,gz,rgrav,w3,pt, &
+!$OMP                                  a_imp,dt,gama,akap,ws,p_fac,scale_m,ms,hs,q_con,cappa) &
+!$OMP                          private(cp2,gm2, dm, dz2, w2, pm2, pe2, pem, peg)
+   do 2000 j=js-1, je+1
+
+      do k=1,km
+         do i=is1, ie1
+            dm(i,k) = delp(i,j,k)
+         enddo
+      enddo
+
+      do i=is1, ie1
+         pef(i,j,1) = ptop                     ! full pressure at top
+         pem(i,1) = ptop
+#ifdef USE_COND
+         peg(i,1) = ptop
+#endif
+      enddo
+
+      do k=2,km+1
+         do i=is1, ie1
+            pem(i,k) = pem(i,k-1) + dm(i,k-1)
+#ifdef USE_COND
+! Excluding contribution from condensates:
+            peg(i,k) = peg(i,k-1) + dm(i,k-1)*(1.-q_con(i,j,k-1))
+#endif
+         enddo
+      enddo
+
+      do k=1,km
+         do i=is1, ie1
+            dz2(i,k) = gz(i,j,k+1) - gz(i,j,k)
+#ifdef USE_COND
+            pm2(i,k) = (peg(i,k+1)-peg(i,k))/log(peg(i,k+1)/peg(i,k))
+
+#ifdef MOIST_CAPPA
+            cp2(i,k) = cappa(i,j,k)
+            gm2(i,k) = 1. / (1.-cp2(i,k))
+#endif
+
+#else
+            pm2(i,k) = dm(i,k)/log(pem(i,k+1)/pem(i,k))
+#endif
+             dm(i,k) = dm(i,k) * rgrav
+             w2(i,k) = w3(i,j,k)
+         enddo
+      enddo
+
+
+      if ( a_imp < -0.01 ) then
+           call SIM3p0_solver(dt, is1, ie1, km, rdgas, gama, akap, pe2, dm, &
+                              pem, w2, dz2, pt(is1:ie1,j,1:km), ws(is1,j), p_fac, scale_m)
+      elseif ( a_imp <= 0.5 ) then
+           call RIM_2D(ms, dt, is1, ie1, km, rdgas, gama, gm2, pe2, &
+                       dm, pm2, w2, dz2, pt(is1:ie1,j,1:km), ws(is1,j), .true.)
+      else
+           call SIM1_solver(dt, is1, ie1, km, rdgas, gama, gm2, cp2, akap, pe2,  &
+                            dm, pm2, pem, w2, dz2, pt(is1:ie1,j,1:km), ws(is1,j), p_fac)
+      endif
+
+      do k=2,km+1
+         do i=is1, ie1
+            pef(i,j,k) = pe2(i,k) + pem(i,k)  ! add hydrostatic full-component
+         enddo
+      enddo
+
+! Compute Height * grav (for p-gradient computation)
+      do i=is1, ie1
+         gz(i,j,km+1) = hs(i,j)
+      enddo
+
+      do k=km,1,-1
+         do i=is1, ie1
+            gz(i,j,k) = gz(i,j,k+1) - dz2(i,k)*grav
+         enddo
+      enddo
+
+2000  continue
+
+  end subroutine Riem_Solver_c
+
+
+!GFDL - This routine will not give absoulte reproducibility when compiled with -fast-transcendentals.
```

Comment:
typo, absolute

Reply:
This will get fixed first in master and then pushed to dev/gfdl to keep it in sync.